keps/973-workload-priority/README.md

```diff
@@ -0,0 +1,349 @@
+`WorkloadPriorityClass` is independent from pod's priority.
+The priority value will be part of the workload spec and be mutable.
+CRDs like Job, MPIJob etc specify the `WorkloadPriorityClass` through labels.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories
+
+Kueue issue [973](https://github.com/kubernetes-sigs/kueue/issues/973) provides details on the initial feature implementation.
+
+#### Story 1
+
+In an organization, admins want to set a lower priority for development workloads and a higher priority for production workloads.
+In such cases, they create two `WorkloadPriorityClass` and apply each one to the respective workloads.
+
+#### Story 2
+
+An organization desires to modify the priority of workloads that remain inactive for a specific duration.
+By developing a custom controller to manage Priority value of `Workload` spec, this expectation can be met.
+
+### Risks and Mitigations
+
+It's possible that the pod's priority conflicts with the workload's priority.
+For example, a high-priority job with low-priority pods may never run to completion because it may always be preempted by kube-scheduler.
+We should document the risks of pod preemption to uses.
+We can also point users to create `PriorityClass` for their pods that are [non-preempting](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class).
+If a workload's priority is high and pod's priority is low and the kube-scheduler initiates preemption, the pod's priority is prioritized. To prevent this behavior, non-preempting setting is needed.
+
+
+## Design Details
+
+### Kueue WorkloadPriorityClass API
+
+We introduce the `WorkloadPriorityClass` API.
+
+```golang
+type WorkloadPriorityClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Value int32   `json:"value,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+```
+
+Also `priorityClassSource` field is added to `WorkloadSpec`.  
+The `PriorityClass` field can accept both k8s `PriorityClass` and `workloadPriorityClass` names as values.
+To distinguish, when using `workloadPriorityClass`, a `priorityClassSource` field has the "kueue.x-k8s.io/WorkloadPriorityClass" value.
+When using k8s `PriorityClass`, a `priorityClassSource` field has the "scheduling.k8s.io/PriorityClass" value.
+
+```golang
+type WorkloadSpec struct {
+  ...
+  PriorityClassSource string `json:"priorityClassSource,omitempty"`
+  ...
+}
+```
+
+### How to use WorkloadPriorityClass on Job
+
+The `workloadPriorityClass` is specified through a label `kueue.x-k8s.io/priority-class`.
```

Comment:
This label is mutable only when the Job isn't yet admitted by ClusterQueue, right?
It would be worth mentioning about conditions to available modify the label.

Reply:
Updated.

Reply:
I think we should just make it mutable.
Note that there is an open KEP to make PriorityClass mutable in k8s

https://github.com/kubernetes/enhancements/pull/4129

Reply:
@Gekko0114 Also, can you add https://github.com/kubernetes/enhancements/pull/4129 to this KEP as an xref?

Reply:
Thanks. Added